torch/testing/_internal/common_device_type.py

```diff
@@ -484,42 +484,48 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
 
         class_name = generic_test_class.__name__ + base.device_type.upper()
 
-        # type set to Any and suppressed due to unsupport runtime class:
-        # https://github.com/python/mypy/wiki/Unsupported-Python-Features
-        device_type_test_class: Any = type(class_name, (base, empty_class), {})
-
-        for name in generic_members:
-            if name in generic_tests:  # Instantiates test member
-                # Requires tests be a function for Python2 compat
-                # (In Python2 tests are type checked methods wrapping functions)
-                test = getattr(generic_test_class, name)
-                if hasattr(test, '__func__'):
-                    test = test.__func__
-                assert inspect.isfunction(test), "Couldn't extract function from '{0}'".format(name)
-
-                # XLA-compat shim (XLA's instantiate_test takes doesn't take generic_cls)
-                sig = inspect.signature(device_type_test_class.instantiate_test)
-                if len(sig.parameters) == 3:
-                    # Instantiates the device-specific tests
-                    device_type_test_class.instantiate_test(name, copy.deepcopy(test), generic_cls=generic_test_class)
-                else:
-                    device_type_test_class.instantiate_test(name, copy.deepcopy(test))
-            else:  # Ports non-test member
-                assert name not in device_type_test_class.__dict__, "Redefinition of directly defined member {0}".format(name)
-
-                # Unwraps to functions (when available) for Python2 compat
-                nontest = getattr(generic_test_class, name)
-                if hasattr(nontest, '__func__'):
-                    nontest = nontest.__func__
-
-                setattr(device_type_test_class, name, nontest)
-
-        # Mimics defining the instantiated class in the caller's file
-        # by setting its module to the given class's and adding
-        # the module to the given scope.
-        # This lets the instantiated class be discovered by unittest.
-        device_type_test_class.__module__ = generic_test_class.__module__
-        scope[class_name] = device_type_test_class
+
+        for is_uncoalesced in set([add_uncoalesced_tests, False]):
```

Comment:
I don't think we want to edit the device generic test framework like this to support toggling coalescing. 

What does this want to do, exactly? Is it that for a test template "test_foo" it wants to instantiate a cross-product of device x dtype x coalesced/uncoalesced, so an instantiated test name would look like "test_foo_cpu_float32_coalesced"?

Reply:
It only instantiates when  `add_uncoalesced_tests` is True if not it remains with the same behavior. 

So, when `add_uncoalesced_tests` is True it generates also the uncoalesced version.  
So in the example two versions will be instantiated:
- "test_foo_cpu_float32"
- "test_foo_cpu_float32_uncoalesced"

Reply:
That makes sense and it's a great goal.

What about adding a decorator, like @coalescedonoff that modifies test generation here:

https://github.com/pytorch/pytorch/blob/789dc6d445136307070486fccf6c9276a722a384/torch/testing/_internal/common_device_type.py#L327

And it could pass it a "is_coalesced" kwarg to the instantiated test function, so the function template signature would become:

```
test_foo(self, device, dtype, *, is_coalesced)
```

and the instantiated functions would have names like test_foo_cuda_float32_coalesced and test_foo_cpu_cfloat_uncoalesced?

A more general approach would be to create a "variants" decorator. It might look like this:

```
@variants(
  ("coalesced", {"is_coalesced : True"}),
  ("uncoalesced", {"is_coalesced : False"})
)
```

That might look like a lot, but with partial this pattern could be saved under a simpler name like @coalescedonoff. Each entry in variants could specify how to extend the test's name, then provide the dict of kwargs and values to pass to the instantiated test.

This is a little like pytest's parameterize functionality, although they support many more options. If you'd like to we could look at some alternative designs based on parameterize's UX.

Would an approach like this make sense? Or should we do something else?

Reply:
Thanks @mruberry, In this PR I was using `is_uncoalesced` as an object attribute for each instance as most of the test methods requires both versions, but looking it in detail, there are some test that don't require uncoalesced tensors. 
So, I agree with this approach.  I think in the meanwhile the @coalescedonoff is enough. I will add this change in a new PR #54572 as it was suggested.

Reply:
SGTM!

Reply:
Note that one other alternative is to do something like this:

https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/torch/testing/_internal/common_cuda.py#L102

Here a new test isn't generated, but the current test is run twice. This is a little hacky, however.